### PR TITLE
Split deserialize and cast

### PR DESCRIPTION
**Pull request by chaymaeBZ**
This fixes compatibility with latest version of [Ruby ULID](https://github.com/rafaelsales/ulid). Therefore we gain the millisecond precision for ulids.

The problem was introduced between Ruby ULID v1.1.1 and v1.2.0. The generated ULID changed encoding from utf8 to ascii-8bit.

In this PR, I split `deserialize` and `cast`. All the logic that was implemented as cast, is needed only for the deserialization of the binary data from db. Documentation about `cast` and `deserialize` [here](https://api.rubyonrails.org/classes/ActiveModel/Type/Value.html#method-i-cast_value).

`assert_valid_value` method is used to validate ULID format compatibility.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -38,6 +38,7 @@ services:
     environment:
       PGDATA: /data
       POSTGRES_DB: db
+      POSTGRES_HOST_AUTH_METHOD: trust
     healthcheck:
       test: echo "\\l" | psql -U postgre
```
Comment (bquorning) [inline docker-compose.yml:43]: To make the tests run locally, I need this change too:
```suggestion
      test: echo "\\l" | psql -U postgres
```

Comment (chaymaeBZ) [reply docker-compose.yml:43]: it works for me just as it is now ..

Comment (bquorning) [reply docker-compose.yml:43]: Could you try this:
1. Start a postgres:12 container:
  `docker run -e PGDATA=/data -e POSTGRES_DB=db -e POSTGRES_HOST_AUTH_METHOD=trust postgres:12`
2. (in another console) connect to the running image and run the `echo|psql` command:
  `docker exec -it $(docker ps | grep postgres:12 | head -n 1 | awk '{ print $1 }') sh -c 'echo "\\l" | psql -U postgre'`

When I do this, I get `psql: error: FATAL:  role "postgre" does not exist`.

When I change to use `psql -U postgres` I get
```
                                 List of databases
   Name    |  Owner   | Encoding |  Collate   |   Ctype    |   Access privileges
-----------+----------+----------+------------+------------+-----------------------
 db        | postgres | UTF8     | en_US.utf8 | en_US.utf8 |
 postgres  | postgres | UTF8     | en_US.utf8 | en_US.utf8 |
 template0 | postgres | UTF8     | en_US.utf8 | en_US.utf8 | =c/postgres          +
           |          |          |            |            | postgres=CTc/postgres
 template1 | postgres | UTF8     | en_US.utf8 | en_US.utf8 | =c/postgres          +
           |          |          |            |            | postgres=CTc/postgres
(4 rows)
```

Comment (chaymaeBZ) [reply docker-compose.yml:43]: That's right! I get the same behaviour following the procedure you mentioned.
Which made me dig deeper into these health checks. They are actually not being used. So we can completely remove them.

Comment (chaymaeBZ) [reply docker-compose.yml]: Thank you @bquorning ! I fixed this `pg12` health check and we no longer rely on `wait-for-it` to do health check. Thus resolving the conversation.

File: test/test_helper.rb [hunk 2]
```diff
@@ -40,27 +42,13 @@
...
 end
 
-ActiveRecord::Base.establish_connection(
-  host: db[:host],
-  adapter: db[:adapter],
-  username: db[:username],
-  password: db[:password],
-  database: db[:name]
-)
+ActiveRecord::Base.establish_connection(db.merge({ database: db[:name] }))
```
Comment (bquorning) [inline test/test_helper.rb:51]: I am not sure why we even have a `name` attribute in the DB config. Can’t we just change it to `database`, then we won’t need this line?

Comment (chaymaeBZ) [reply test/test_helper.rb]: Well that is right! but we can't establish a connection to a db that's not created yet (when using mysql for example). but I have updated the code now. slicing only relevant attributes for establishing connection before db creation 💪🏼

File: test/test_helper.rb [hunk 3]
```diff
@@ -7,60 +7,48 @@
...
-if db[:adapter] != 'sqlite3'
-  ActiveRecord::Base.establish_connection(
-    host: db[:host],
-    adapter: db[:adapter],
-    username: db[:username],
-    password: db[:password],
-  #  database: db[:name]
-  )
+unless db[:adapter] == 'sqlite3'
+  ActiveRecord::Base.establish_connection(db.slice(:host, :username, :password, :adapter))
```
Comment (bquorning) [inline test/test_helper.rb:45]: ActiveSupport adds some syntactic sugar to Hash:
```suggestion
  ActiveRecord::Base.establish_connection(db.except(:database))
```

Comment (chaymaeBZ) [reply test/test_helper.rb]: moved to #5

**Review by bquorning — APPROVED** [review]

**Pull request merged**